### PR TITLE
Block potentially misbehaving peers

### PR DESCRIPTION
**Pull request by pradovic**
https://github.com/ethersphere/bee/issues/152

### CODE REVIEW

**Review by janos — APPROVED** [review]
👍

File: pkg/p2p/error.go [hunk 1]
```diff
@@ -65,6 +65,33 @@ func (e *DisconnectError) Error() string {
...
 }
 
+type BlockPeerError struct {
+	duration time.Duration
+	err      error
+}
+
+// NewBlockPeerError wraps error and creates a special error that is treated specially
+// by p2p. It causes peer to be disconnected and blocks any new connection for this peer for the provided duration.
+//
```
Comment (acud) [inline pkg/p2p/error.go:75]: can probable be removed

File: pkg/p2p/error.go [hunk 2]
```diff
@@ -65,6 +65,33 @@ func (e *DisconnectError) Error() string {
...
+
+// Unwrap returns an underlying error.
+func (e *BlockPeerError) Unwrap() error { return e.err }
+
+// Error implements function of the standard go error interface.
+func (e *BlockPeerError) Error() string {
+	return e.err.Error()
+}
+
+func (e *BlockPeerError) Duration() time.Duration {
```
Comment (acud) [inline pkg/p2p/error.go:91]: add godoc?

File: pkg/p2p/libp2p/internal/blocklist/blocklist_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package blocklist_test
```
Comment (acud) [inline pkg/p2p/libp2p/internal/blocklist/blocklist_test.go:1]: license missing

File: pkg/p2p/libp2p/internal/blocklist/export_test.go [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package blocklist
```
Comment (acud) [inline pkg/p2p/libp2p/internal/blocklist/export_test.go:1]: ditto

File: pkg/p2p/libp2p/libp2p.go [hunk 5]
```diff
@@ -223,6 +228,20 @@ func New(ctx context.Context, signer beecrypto.Signer, networkID uint64, overlay
 			return
 		}
 
+		blocked, err := s.blocklist.Exists(i.BzzAddress.Overlay)
+		if err != nil {
+			s.logger.Debugf("blocklisting: eixsts %s: %v", peerID, err)
```
Comment (acud) [inline pkg/p2p/libp2p/libp2p.go:233]: `eixsts` -> `exists`

File: pkg/p2p/libp2p/libp2p.go [hunk 6]
```diff
@@ -311,8 +331,20 @@ func (s *Service) AddProtocol(p p2p.ProtocolSpec) (err error) {
...
+
+				var bpe *p2p.BlockPeerError
+				if errors.As(err, &bpe) {
+					if err := s.blocklist.Add(overlay, bpe.Duration()); err != nil {
+						s.logger.Debugf("blocklist: could blocklist peer %s: %v", peerID, err)
+						s.logger.Errorf("unable to blocklist peer %v", peerID)
+						_ = s.Disconnect(overlay)
+					}
+
+					s.logger.Trace("blocklistted a peer %s", peerID)
```
Comment (acud) [inline pkg/p2p/libp2p/libp2p.go:347]: blocklisted

File: pkg/p2p/libp2p/libp2p.go [hunk 7]
```diff
@@ -383,6 +426,20 @@ func (s *Service) Connect(ctx context.Context, addr ma.Multiaddr) (address *bzz.
 		return nil, fmt.Errorf("handshake: %w", err)
 	}
 
+	blocked, err := s.blocklist.Exists(i.BzzAddress.Overlay)
+	if err != nil {
+		s.logger.Debugf("blocklisting: eixsts %s: %v", info.ID, err)
```
Comment (acud) [inline pkg/p2p/libp2p/libp2p.go:431]: exists

File: pkg/p2p/libp2p/libp2p.go [hunk 8]
```diff
@@ -383,6 +426,20 @@ func (s *Service) Connect(ctx context.Context, addr ma.Multiaddr) (address *bzz.
...
+		s.logger.Errorf("internal error while connecting with peer %s", info.ID)
+		_ = s.disconnect(info.ID)
+		return nil, fmt.Errorf("peer blocklistted")
```
Comment (acud) [inline pkg/p2p/libp2p/libp2p.go:434]: blocklisted

File: pkg/p2p/libp2p/libp2p.go [hunk 9]
```diff
@@ -383,6 +426,20 @@ func (s *Service) Connect(ctx context.Context, addr ma.Multiaddr) (address *bzz.
...
+	}
+
+	if blocked {
+		s.logger.Errorf("blocked connection from blocklisted peer %s", info.ID)
+		_ = s.disconnect(info.ID)
+		return nil, fmt.Errorf("peer blocklistted")
```
Comment (acud) [inline pkg/p2p/libp2p/libp2p.go:440]: blocklisted

**Review by acud — APPROVED** [review]
minor cosmetics. otherwise LGTM :+1:

**Comment by pradovic** [comment]
Thanks, @acud, fixed!

**Review by acud — APPROVED** [review]

**Pull request merged**